tests/samples_test.py

```diff
@@ -0,0 +1,448 @@
+        ),
+        (
+            smolyay.samples.NestedClenshawCurtisPointSet([-1, 1], 3),
+            3,
+            [
+                -numpy.sqrt(numpy.sqrt(2) + 1) / (2**0.75),
+                -numpy.sqrt(numpy.sqrt(2) - 1) / (2**0.75),
+                numpy.sqrt(numpy.sqrt(2) - 1) / (2**0.75),
+                numpy.sqrt(numpy.sqrt(2) + 1) / (2**0.75),
+            ],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            0,
+            [0]),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            1,
+            [-1, 1]),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            2,
+            [-1 / numpy.sqrt(2), 1 / numpy.sqrt(2)],
+        ),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            3,
+            [
+                -numpy.sqrt(numpy.sqrt(2) + 1) / (2**0.75),
+                -numpy.sqrt(numpy.sqrt(2) - 1) / (2**0.75),
+                numpy.sqrt(numpy.sqrt(2) - 1) / (2**0.75),
+                numpy.sqrt(numpy.sqrt(2) + 1) / (2**0.75),
+            ],
+        ),
+        (smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5), 4, []),
+        (
+            smolyay.samples.SlowNestedClenshawCurtisPointSet([-1, 1], 5),
+            5,
+            [
+                -numpy.sqrt(2 + numpy.sqrt(2 + numpy.sqrt(2))) / 2,
+                -numpy.sqrt(2 + numpy.sqrt(2 - numpy.sqrt(2))) / 2,
+                -numpy.sqrt(2 - numpy.sqrt(2 - numpy.sqrt(2))) / 2,
+                -numpy.sqrt(2 - numpy.sqrt(2 + numpy.sqrt(2))) / 2,
+                numpy.sqrt(2 - numpy.sqrt(2 + numpy.sqrt(2))) / 2,
+                numpy.sqrt(2 - numpy.sqrt(2 - numpy.sqrt(2))) / 2,
+                numpy.sqrt(2 + numpy.sqrt(2 - numpy.sqrt(2))) / 2,
+                numpy.sqrt(2 + numpy.sqrt(2 + numpy.sqrt(2))) / 2,
+            ],
+        ),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2),
+            0,
+            [0]),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2),
+            1,
+            [2 * numpy.pi / 3, 4 * numpy.pi / 3],
+        ),
+        (
+            smolyay.samples.NestedTrigonometricPointSet([0, 2 * numpy.pi], 2),
+            2,
+            2
+            * numpy.pi
+            * numpy.array(
+                [
+                    1 / 9,
+                    2 / 9,
+                    4 / 9,
+                    5 / 9,
+                    7 / 9,
+                    8 / 9,
+                ]
+            ),
+        ),
+    ],
+    ids=nested_sample_ids,
+)
+def test_nested_get_individual_levels(nested_samples, level, points):
+    """test number of points per level"""
+    end_lev = nested_samples.end_level[level]
+    start_lev = nested_samples.start_level[level]
+    assert numpy.allclose(nested_samples.points[start_lev:end_lev], points)
```

Comment:
I didn't catch this before, but what is actually the purpose of this test? You already have a test that checks that the points are generated correctly for each type of point set, and you also have a test that makes sure the nested levels are defined correctly. Hence, I think all this is doing is making sure that Python's slice operator works, which we don't need to test. I would remove this.

Reply:
That is true that it's testing python's slice operator, but the main purpose is to check that what I think the `start_level` and `end_level` are supposed to be actually results in the desired behavior. It would very bad if what I or someone else assumed would get the right points in each level actually didn't, and the `start_level` and `end_level` calculations were fitted to match something that wasn't right. It essentially tests both the class and the answers to the previous test.

Reply:
> It essentially tests both the class and the answers to the previous test.

If this is the only purpose of this test, it should be removed because it duplicates the others.

Reply:
> If this is the only purpose of this test, it should be removed because it duplicates the others.

So we should either test that the `start_level` and `end_level` equal some predetermined lists or test that `samples.points[samples.start_level[i] : samples.end_level[i]]` nets you the expected points in level `i`, but not both?

If so, I would actually prefer to do the latter than the former.

Reply:
> So we should either test that the `start_level` and `end_level` equal some predetermined lists

You have already implemented this test, it's `test_nested_levels`. And `test_generate_points` checks that the points are created in the right order. So, there is no need to check slicing a list of points (that are known to be correct) between two indices (that are known to be correct).

Reply:
I've already implemented both checks, so I can remove either. I would prefer keeping the test where there's no risk of me or anyone else making an off-by-one error, and that's the test that checks that `start_level` and `end_level` get the right slicing behavior and not the test that checks if they equal certain lists.

Reply:
okay